### ISSUE TITLE
What's the best way to handle $repeat updates when each repeated object can change individually?

### ISSUE THREAD

**Issue by pitus** [Polymer/lit-html#361]
Here's the scenario I'm facing (having some trouble with it, coming from the knockout.js `observableArray` realm):
1) Want to show a list of Users as `tr` rows of a table
2) Can't just add a User element to generate a `tr` so I have to just map my users to User class instances (unless there's another way to output `<tr><...></tr>` for an element I'm not aware of?
3) Once `$repeat`ed I get the desired output, however **each User's property may change due to an action taken** - such as reloading some of user's data via a button press.
4) How should I re-trigger `_render` when any `User` changes? dispatch a `CustomEvent` from each User? If so, how can I do it if User is just a regular class and not an element in the DOM? Also, what's the best way to listen to all of them in order to re-trigger `_render` ? `addEventListener` for my custom event? I'm assuming that there's no way to have `_render` automatically detect each User's change as it can only monitor host's properties?
5) If I `_requestRender` once I get notified of a User change, it will re-render the entire list of Users, will it not? **How can this be optimized so that only the changed User template is updated** (that's the whole idea behind Lit - minimize updates and limit to only elements that have changed) ?

**Comment by pitus** [comment]
BTW, I was thinking of how to generate an HTML TABLE ROW per User element while User is based on the LitElement ... but I can't figure out how to have it render `<tr><td ...></tr>` as it always renders my custom component tag with its own ShadowDom being `<tr>...</tr>` but Chrome doesn't like my custom element as a child of TABLE or TBODY.
Is there any way to attach my User's custom element's shadow root to a parent TR instead? Then I could generate the TR via parent template (say Users element) and then attach user element to that TR to manage its TDs as shadow dom? This way, User element could handle its own render on any property change as needed, optimized (only changes updated).

**Comment by pitus** [comment]
Nope, can't `attachShadow` to TR
So, how can I have my custom element 'skip' the custom element tag in HTML and replace it with a TR or add itself as multiple TDs of a parent TR or somehow re-use a parent template with a TR that includes placeholders for each `User` property - sort of like inverse `slot`s ?
In general, this is a case of how a single custom element might 'populate' multiple places of its parent? Slots allow the opposite - the parent to populate multiple slots of a child element's template.

I may have to review in more detail how render handles templates via `html` literal function to see how it updates an existing element with a new value and maybe I'll be able to figure out another way to manage sub/nested template updates using a regular class? Any thoughts / suggestions on that?

**Comment by pitus** [comment]
OK, I give up studying the lit-html source code, it's way too much to figure out on my own right now :(
Perhaps there's some kind of a developer documentation on how it works? I thought that the browser parses literal template and splits it into strings + values and these are then efficiently updated, but it looks like there's a lot of 'magic' behind the scenes with markers, tree walkers ... is it really that efficient with regex matching all over the place?

Oh, and I still don't know how to build a simple table with User objects within it :(
Is there any way to have a regular class manage a template that's then used by parent's render somehow?

**Comment by merlinnot** [comment]
Maybe you could just use a CSS grid instead of a table?

**Comment by pitus** [comment]
@merlinnot - I need this to resize columns based on the content, so the table is the easiest html syntax to do this. With CSS, I'll have to worry about each row's widest value in the table and try to adjust column widths to match that myself - unless there's a way to have CSS do it all for you for all columns, not just the single auto column? I'll have to spend some time experimenting with it I suppose if there's no other way to do this :(

**Comment by pitus** [comment]
For now I'm investigating how directives are implemented to see if, maybe I can use that instead of a custom element (which can't be done for the `tr`)

**Comment by Lodin** [comment]
@pitus You definitely need [CSS Grid](https://css-tricks.com/snippets/css/complete-guide-grid/) as @merlinnot mentioned. AFAIK, it allows to do everything `table` does but without using the table itself. 

Actually, using `table` with web components is one of the problems came with web components. All implementations I've seen dropped support for native table and ended up using own implementation based on web components. E.g., [vaadin-grid](https://www.webcomponents.org/element/vaadin/vaadin-grid). 

So, I suppose it would be good to take a look at CSS Grid. It came to solve different problems with `table`.

**Comment by justinfagnani** [comment]
@pitus I'm having a very hard time following the actual use case here. Do you have some example code of what you're trying to do?

A lot of the questions also seem to have to do with LitElement, not lit-html. lit-html doesn't have any `_render()`, or `_requestRender()` methods.

**Comment by PaulHMason** [comment]
I'm not quite sure what you mean, but I think I did something similar. I used the CSS table display types rather than the table element types (rows, cells and tables being their own custom lit-elements so that their rendering can be managed at the various levels of detail). That should work fine unless you need anything the CSS types don't support (like column spans).

**Comment by merlinnot** [comment]
There is a way to do it in a grid, fairy simple one. Take a look at `minmax`, `max-content`, `min-content` and pick what fits you best. Unless I didn’t understand your needs correctly, but then you can just take a look at other things in the docs :)

**Comment by pitus** [comment]
Before I'll prepare a use case for @justinfagnani of what I'm trying to achieve, I'll first try to use either grid CSS types (suggested by @PaulHMason) or switch completely to CSS grid (as suggested by @merlinnot and @Lodin) and see how that comes out. Thanks, I'll post my results when/if I succeed.

**Comment by pitus** [comment]
Actually, the problem with `<td>` not working under custom element, forcing me to use `<div class="td">` was probably just an issue with [jsfiddle](http://jsfiddle.net/1djs97yv/182/) - see how `Row 1 Column 1`, which uses a regular `<td>`, is shown below the table...
Turns out that in my own version, I was able to just use regular `<td>`s even with colspan, as children of my own custom element (shadow DOM), very cool ;)

**Comment by pitus** [comment]
I'd like to re-open this subject as I'm still struggling with the `repeat` directive.
Specifically, I'm iterating over an array of Objects and displaying their properties (building a generic Grid component that uses a meta definition for columns to map to Object's data) in a table.
My problem is that each Object fed to `repeat` may have its properties changed either by user's input or by some background operation such as fetching data asynchronously...
What's the 'correct' way to have that table row (or repeated template) refresh (only for that single Object that has changed, not the entire repeated set) when Object's property changes? I can dispatch some kind of an event from the Object when it changes and have the 'grid' component react to it by calling `_requestRender()` but I'm not sure how to refresh just a single entry/template created by the `repeat` directive?
For now, I've created a new `grid-row` component that I'm using instead of a regular `tr` and have it listen to the `change` event and call it's own `_requestRender()` but that's way too much overhead for such a seemingly simple task.
Is there a better way now or planned in the future to support some kind of a notification mechanism or 'observable' pattern in lit-element ?

While I'm on the subject of `repeat` - I need some simple access to each 'row' rendered so that I can control its `classList` based on Object rendered (its properties). I could try adding extra properties to each Object defining what css class should be rendered for the row but I'd much rather just access the HTMLelement for the row and directly change its `classList`. The problem is that I'd have to define an id for each row being `repeat`ed and then 'find' it by id to access ... and again, that seems like a lot of overhead. I'm coming from the knockout library where I could access each model and element easily, without assigning any ids. For now, I'm setting Object._row property in my custom grid-row component when it's already been rendered ... but that's such an ugly hack, I'd rather find a better solution soon ;)

**Comment by thescientist13** [comment]
@pitus 
I think I may be in a similar situation.  Maybe I'm just struggling a bit with understanding the nuances of attributes vs properties and how change detection is handled in **lit-html**?

Basically, say I have a `ListComponent` (your `<table>`).  it renders _N_ `ListItemComponent`s (your `<tr>`).  In the same manner, my thought was just to do all the sorting in the `ListComponent`, and then have each list item just take a `todo` object and render.  However, this doesn't seem to trigger change detection, say when one of the properties changes (`isCompleted`)


```javascript
        <ol>
          ${repeat(todos, (todo) => todo.id, (todo) => html`
            <pre>${JSON.stringify(todo)}</pre> /* {"completed":true,"task":"sss","id":1534082067216,"created":1534082067216} */
            <li>
              <x-todo-list-item 
                todo=${todo}
              ></x-todo-list-item>
            </li>
          `)}
        </ol>
```

I know the custom events are working, because counters in the `ListComponent` are updating based on `this.todos.length` and `completed` switches between `true` / `false` correctly.  `ListItemComponent` just happens to be the only component observing an `Object`.  

In this screenshot for example, the ⛔️ should be a ✅ 
![screen shot 2018-08-12 at 10 00 43 am](https://user-images.githubusercontent.com/895923/44002884-167ff40c-9e18-11e8-8712-4733c96eb5f6.png)


Now, I am ok being told this is the wrong way to do this, haha.  For instance, if I do something like this, it seems to work as expected (⛔️ turns into a ✅)

```javascript
        <ol>
          ${repeat(todos, (todo) => todo.id, (todo) => html`
            <pre>${JSON.stringify(todo)}</pre> /* {"completed":true,"task":"sss","id":1534082067216,"created":1534082067216} */
            <li>
              <x-todo-list-item 
                completed=${todo.completed}
                title=${todo.task}
              ></x-todo-list-item>
            </li>
          `)}
        </ol>
```

So yeah, I'm just as curious on if there are pro tips on the best way to pass data / handle nested components / etc.  I understand the challenges of observing deeply nested objects, so just curious what the good use cases are for passing arrays / objects (if any) and / or if this is a better use case for state management or whatnot.

**Comment by ruphin** [comment]
The reason why it works with `() => Date.now()` is that lit-html will consider every element in the list to be unique, so it will regenerate the whole list every time. This completely defeats the purpose of using `repeat()`

The idea of `repeat()` is that it understands the relationship between content in different renders. This is to fix a common problem when rendering Arrays:


If you tell lit-html to render an Array, it will simply render the content without considering what it is or where it came from, every time. That means that it does not account for changes made to the nodes in between renders, and just renders in the new content. If you make any changes to any nodes in between, those changes will not 'move with the nodes'. Here is an example:

https://codepen.io/ruphin/pen/pZBoRR?editors=0010

That renders the list [1,2,3,4,5], changes the color of the first element (the 1) to red, and then renders the list [5,4,3,2,1]. The expected behaviour would be for the 1 to stay rendered in red, but the actual behaviour is that the first element stays red (which is now the 5)

`repeat()` solves this by making it possible for lit-html to understand the relationship of items we are rendering. Basically, if you try to render the same item again, it will not create a new node, but instead move the item that we previously rendered to the right place. The key function is there to help lit understand which items are 'the same'.

https://codepen.io/ruphin/pen/xJexLJ?editors=0010

Now you see that instead of rendering the list again, it only shuffles around the nodes, which keeps the 1 in red. The side-effect of this is that if the key is the same, it does not render the item again, which caused your original issue. Your solution of providing `Date.now()` as a key basically clobbers any reference to anything you had rendered before, forcing lit to render every item again. This is probably the worst of all possible solutions, since it does not cache the templates, nor is it able to re-use any rendered nodes.

In short:

Render arrays if you are rendering 'pure' items that do not change their own state, and all the state is passed down by the render template. You cannot ever change the state of the rendered items in any way other than rendering them again, or you get the red 1 problem from the first example.

Render with `repeat` if you are rendering items that manage their own state and have no state passed down. Once the item is rendered it has to manage it's own state completely, since existing items will never be rendered again.

**Comment by thescientist13** [comment]
Thank you for the clear and detailed reply @ruphin , and makes sense!

> The reason why it works with () => Date.now() is that lit-html will consider every element in the list to be unique, so it will regenerate the whole list every time.

Yeah, this was my hunch, in that because **lit-html** couldn't detect changes of the underlying object properties, blowing away the template reference was essentially a brute force approach, which as you point out, negates the benefits of using `html`.

Really appreciate your time, and for clearing this up for me. 

One question re: this
>  Render arrays if you are rendering 'pure' items that do not change their own state ... You cannot ever change the state of the rendered items in any way _other than rendering them again._

So for rendering `Array`s, you mean like this?
```javascript
<ol>
  ${todos.map((todo) => { 
    return html`
      <li>
        <x-todo-list-item 
          todo=${todo}
        ></x-todo-list-item>
       </li>
    `;
  })}
</ol>
```

And If using **lit-element**, do you mean by calling `_render()` manually after updating the Array and passing in the updated Array?
```javascript
this.addTodo(todo) {
  // push the new array
  
  this.render(this.todos);
}
```

**Comment by ruphin** [comment]
That would be an example of rendering an Array, yes.

Your solution for rendering in **lit-element** works. There are probably multiple ways in **lit-element** to handle Arrays as source datatypes, but I'm not a huge fan of **lit-element** so I'm probably not the right person to tell you what to do here.

**Comment by jfrazzano** [comment]
Do you have control over/ access to userObject.prototype?

**Comment by justinfagnani** [comment]
I think the topic has been covered here pretty well. The main upshot is that like any other one-way, top-down, non-observing, rendering system, notifications of updates and triggers to re-render have to happen out of band. There are a large number of ways to approach this, and hopefully we'll find a way to point people in the right directions in the docs.
